### PR TITLE
#2288 - Load test gateway devops and local docker

### PR DESCRIPTION
**Pull request by dheepak-aot**
- [x] Created deploy.yml and update make to deploy the `load-test-gateway` in openshift
- [x] Created local make and docker compose for load test gateway
image

image

### CODE REVIEW

File: devops/openshift/load-test-gateway-deploy.yml [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+    metadata:
+      name: ${NAME}-hpa
+    spec:
+      scaleTargetRef:
+        apiVersion: apps/v1
+        kind: ReplicationController
+        name: ${NAME}-rc
+      minReplicas: "${{REPLICAS}}"
+      maxReplicas: 10
+      targetCPUUtilizationPercentage: 80
```
Comment (andrepestana-aot) [inline devops/openshift/load-test-gateway-deploy.yml:124]: Do we need autoscaling for this one?

**Review by andrepestana-aot — COMMENTED** [review]
Great job! Just left a comment.

File: devops/openshift/load-test-gateway-deploy.yml (hunk 1 above)
Comment (andrewsignori-aot) [reply devops/openshift/load-test-gateway-deploy.yml:124]: I would recommend not using it. Before the test, it would be good to have them scaled to what we believe would be best to avoid new pods creation.

**Comment by andrepestana-aot** [comment]
Please check the sonar issues.

File: devops/Makefile [hunk 2]
```diff
@@ -433,6 +449,32 @@ deploy-queue-consumers:
...
+	test -n "$(BUILD_NAMESPACE)"
+	test -n "$(BUILD_REF)"
+	test -n "$(LOAD_TEST_GATEWAY_BUILD_REF)"
+	test -n "$(DB_SERVICE_NAME)"
+	test -n "$(LOAD_TEST_GATEWAY_SECRET_NAME)"
+	test -n "$(LOAD_TEST_GATEWAY)"
+	test -n "$(LOAD_TEST_GATEWAY_PORT)"
+	test -n "$(KEYCLOAK_AUTH_URL)"
+	test -n "$(KEYCLOAK_REALM)"
+	@echo "+\n++ Deploying Queues with tag: $(BUILD_REF)\n+"
```
Comment (andrewsignori-aot) [inline devops/Makefile:463]: Please review the comment "Deploying Queues with tag".

File: devops/Makefile [hunk 3]
```diff
@@ -433,6 +449,32 @@ deploy-queue-consumers:
...
+	@oc -n $(NAMESPACE) process -f openshift/queue-consumers-deploy.yml \
```
Comment (andrewsignori-aot) [inline devops/Makefile:464]: Please review the command "openshift/queue-consumers-deploy.yml".

File: devops/openshift/load-test-gateway-deploy.yml [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+                  valueFrom:
+                    secretKeyRef:
+                      key: ${DB_PASSWORD_KEY}
+                      name: ${DB_SECRET_NAME}
+                - name: POSTGRES_DB
+                  valueFrom:
+                    secretKeyRef:
+                      key: ${SIMS_DB_NAME}
+                      name: ${LOAD_TEST_GATEWAY_SECRET_NAME}
+                - name: DISABLE_ORM_CACHE
```
Comment (andrewsignori-aot) [inline devops/openshift/load-test-gateway-deploy.yml:54]: Is `DISABLE_ORM_CACHE` needed for the `load-test-gateway`?

File: sources/Makefile [hunk 5]
```diff
@@ -201,6 +202,12 @@ queue-consumers:
 	@docker-compose -f docker-compose.yml build queue-consumers
 	@docker-compose -f docker-compose.yml up -d queue-consumers
 
+# Runs load test gateway application
+load-test-gateway:
+	@echo "+\n++ Make: Queue only"
```
Comment (andrewsignori-aot) [inline sources/Makefile:207]: Please review the comment.

File: devops/openshift/load-test-gateway-deploy.yml (hunk 4 above)
Comment (dheepak-aot) [reply devops/openshift/load-test-gateway-deploy.yml:54]: Not necessarily, but I felt it is better to ignore the orm cache part if we are not providing redis context to the application

File: sources/packages/backend/apps/load-test-gateway/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+# Installing dependencies
+RUN npm ci
+
+# Copying sources.
+COPY ./apps/load-test-gateway ./apps/load-test-gateway
+COPY ./libs ./libs
+
+# Building app
+RUN npm run build load-test-gateway
+
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/load-test-gateway/Dockerfile:18]: Are we missing `EXPOSE ${PORT}` here or is there a reason to not have it added?

File: sources/packages/backend/apps/load-test-gateway/Dockerfile.dev [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+# Copying root files to image workdir.
+COPY /*.json /*.js ./
+
+# Installing dependencies
+RUN npm ci
+
+# Copying sources.
+COPY ./apps/load-test-gateway ./apps/load-test-gateway
+COPY ./apps/db-migrations ./apps/db-migrations
```
Comment (andrewsignori-aot) [inline sources/packages/backend/apps/load-test-gateway/Dockerfile.dev:22]: Do we need migrations on this POD? Is there a benefit?

File: sources/packages/load-test/src/_tests_/workflow/workflow-assessment-submission.test.ts [hunk 8]
```diff
@@ -14,7 +14,7 @@ import { ClientSecretCredential } from "../../utils/auth";
 /**
  * Load test number of iterations to run.
  */
-const ITERATIONS = 500;
+const ITERATIONS = 15;
```
Comment (andrewsignori-aot) [inline sources/packages/load-test/src/_tests_/workflow/workflow-assessment-submission.test.ts]: Is this change on purpose to keep the number low for now?

**Review by andrewsignori-aot — COMMENTED** [review]
Thanks for bringing this into the current effort. Please take a look at the comments.
I believe that we need a ticket to have the GitHub actions available to do the DEV/TEST build deploy.

File: sources/packages/load-test/src/_tests_/workflow/workflow-assessment-submission.test.ts (hunk 8 above)
Comment (dheepak-aot) [reply sources/packages/load-test/src/_tests_/workflow/workflow-assessment-submission.test.ts]: It was not intentional 😊

File: sources/packages/backend/apps/load-test-gateway/Dockerfile.dev (hunk 7 above)
Comment (dheepak-aot) [reply sources/packages/backend/apps/load-test-gateway/Dockerfile.dev:22]: It is not required and there is no benefit. The boiler plate code came from queue-consumers and this one missed to be removed.

File: sources/packages/backend/apps/load-test-gateway/Dockerfile (hunk 6 above)
Comment (dheepak-aot) [reply sources/packages/backend/apps/load-test-gateway/Dockerfile:18]: Correct me if wrong, I am not sure about the impact of adding expose, as queue-consumers(bull dashboard) is working without the expose and I was assuming the same for here as well. may be I am missing something

Comment (andrewsignori-aot) [reply sources/packages/backend/apps/load-test-gateway/Dockerfile:18]: As we checked, the `yaml` has the port expose 😉

**Review by andrewsignori-aot — APPROVED** [review]
Thanks again for extending the work and making the DevOps part also compled in the ticket. Look good 👍

File: devops/Makefile [hunk 9]
```diff
@@ -433,6 +449,32 @@ deploy-queue-consumers:
 		| oc -n $(NAMESPACE) apply -f -
 	$(call rollout_and_wait,dc/$(QUEUE_CONSUMERS))
 
+deploy-load-test-gateway:
```
Comment (guru-aot) [inline devops/Makefile:452]: Just for discussion, if we use this container only for load-testing, can this be used for db-migrations too? If we properly spin this up and adding the db-migrations in the Dockerfile?

**Review by ann-aot — APPROVED** [review]
LGTM @dheepak-aot 👍

File: devops/Makefile (hunk 9 above)
Comment (dheepak-aot) [reply devops/Makefile:452]: I believe there is no intention to use this container for db migrations.

Comment (andrewsignori-aot) [reply devops/Makefile:452]: This container may be deployed only to DEV, it may be not present always, and I am struggling to see a benefit in doing it. Can you please provide a benefit for doing it?

**Review by andrepestana-aot — APPROVED** [review]
Looks good! Great job!

**Review by sh16011993 — APPROVED** [review]
Nice work @dheepak-aot  👍

**Review by guru-aot — APPROVED** [review]
LGTM, nice work @dheepak-aot

File: devops/Makefile (hunk 9 above)
Comment (guru-aot) [reply devops/Makefile:452]: This was a temporary container used for load-test, so I thought instead of having 2 jobs, if we extend it for db-migrations, we can use it during revert. I am aware that this may be only deployed in DEV, in case if we extend its functionalities as a common container, I was suggesting.
